src/Engine/EmitMSIL/CodeGenIL.cs

```diff
@@ -706,11 +726,160 @@ private void EmitLanguageBlockNode(AssociativeNode node)
+            {
+                methodName = doubleRangeMethodName;
+            }
+
+            //we still have not selected a method, lets check if our inputs are idents and have known types.
+            if(methodName == unselectedToken)
+            {
+                //if we are generating a simple range and we have all doubles or ints we know what methods to call
+                //in other cases we can't determine which overload to call without the values of these idents.
+                if (stepOp == DSASM.RangeStepOperator.StepSize)
+                {
+                    if (new[] { fromNode, toNode, stepNode }.All(x => x is IdentifierNode ident && CheckIdentType<long>(ident))){
+                        methodName = intRangeMethodName;
+                    }
+                    else if (new[] { fromNode, toNode, stepNode }.All(x => x is IdentifierNode ident && CheckIdentType<double>(ident)))
+                    {
+                        methodName = doubleRangeMethodName;
+                    }
+                }
+            }
+            if (methodName == unselectedToken)
+            {
+                //if we still have not been able to determine the type of range to generate, temporarily generate a double range.
+                //TODO when we add alphabetic ranges we'll want to check for string/char types in the above logic but,
+                //if we still get to this line we'll need to call a dynamic version of generate range that boxes objects.
+                methodName = doubleRangeMethodName;
+            }
+               
+            //call the generate method we've selected.
+
+            IntNode op = null;
+            switch (stepOp)
+            {
+                case DSASM.RangeStepOperator.StepSize:
+                    op = new IntNode(0);
+                    break;
+                case DSASM.RangeStepOperator.Number:
+                    op = new IntNode(1);
+                    break;
+                case DSASM.RangeStepOperator.ApproximateSize:
+                    op = new IntNode(2);
+                    break;
+                default:
+                    op = new IntNode(-1);
+                    break;
+            }
+            var arguments = new List<AssociativeNode>
+            {
+                fromNode,
+                toNode,
+                stepNode ?? new NullNode(),
+                op,
+                AstFactory.BuildBooleanNode(stepNode != null),
+                AstFactory.BuildBooleanNode(hasAmountOperator),
+            };
+
+            
+            var rangeExprFunc = AstFactory.BuildFunctionCall(methodName, arguments);
+            var idlist = new IdentifierListNode()
+            {
+                LeftNode = new IdentifierNode(typeof(Builtin.RangeHelpers).FullName),
+                RightNode = rangeExprFunc
+            };
+            //we want to cache the call to generate range, so traverse down in any case.
+            var t = DfsTraverse(idlist);
```

Comment:
Are the cached generate range methods ever going to be used after retrieving them from the cache?

Reply:
Does this mean, that the range methods will be looked up only the first time a range expr is encountered and subsequently they will simply be returned from the cache? I think I got it.

Reply:
yes, sorry I should probably remove this comment, it was from before when the methodlookup phase made it down here.